src/plugins/profiling/server/routes/stacktrace.ts

```diff
@@ -23,36 +23,98 @@ import { getHitsItems, getDocs } from './compat';
+  // +----------------+--------+----------------+--------+----
+  // |     File ID    |  Addr  |     File ID    |  Addr  |
+  // +----------------+--------+----------------+--------+----
+  FrameID: string;
+
+  // This field is a run-length encoding of a list of uint8s. The order is
+  // reversed from the original input.
+  Type: string;
 }
 
-// extractFileIDArrayFromFrameIDArray extracts all FileIDs from the array of FrameIDs
-// and returns them as an array of base64url encoded strings. The order of this array
-// corresponds to the order of the input array.
-function extractFileIDArrayFromFrameIDArray(frameIDs: string[]): string[] {
-  const fileIDs = Array<string>(frameIDs.length);
-  for (let i = 0; i < frameIDs.length; i++) {
-    fileIDs[i] = extractFileIDFromFrameID(frameIDs[i]);
+// runLengthDecodeReverse decodes a run-length encoding for the reversed input array.
+//
+// The input is a binary stream of 2-byte pairs (first byte is the length and the
+// second byte is the binary representation of the object). The output is a list of
+// uint8s in reverse order.
+//
+// E.g. byte array [5, 0, 2, 2] is converted into an uint8 array like
+// [2, 2, 0, 0, 0, 0, 0].
+export function runLengthDecodeReverse(input: Buffer, outputSize?: number): number[] {
+  let size;
+
+  if (typeof outputSize === 'undefined') {
+    size = 0;
+    for (let i = 0; i < input.length; i += 2) {
+      size += input[i];
+    }
+  } else {
+    size = outputSize;
   }
-  return fileIDs;
+
+  const output: number[] = new Array(size);
+
+  let idx = 0;
+  for (let i = input.length - 1; i >= 1; i -= 2) {
+    for (let j = 0; j < input[i - 1]; j++) {
+      output[idx] = input[i];
+      idx++;
+    }
+  }
+
+  return output;
+}
+
+// decodeStackTrace unpacks an encoded stack trace from Elasticsearch
+export function decodeStackTrace(input: EncodedStackTrace): StackTrace {
+  const countsFrameIDs = input.FrameID.length / BASE64_FRAME_ID_LENGTH;
+  const fileIDs: string[] = new Array(countsFrameIDs);
+  const frameIDs: string[] = new Array(countsFrameIDs);
+
+  // Step 1: Convert the base64-encoded frameID list into two separate
+  // lists (frame IDs and file IDs). The first 22 bytes of a frame ID
+  // contains the FileID.
+  for (let i = 0; i < input.FrameID.length; i += BASE64_FRAME_ID_LENGTH) {
+    const frameID = input.FrameID.slice(i, i + BASE64_FRAME_ID_LENGTH);
+    const fileID = frameIDToFileIDCache.get(frameID) as string;
+    const j = Math.floor(i / BASE64_FRAME_ID_LENGTH);
+
+    frameIDs[j] = frameID;
+
+    if (fileID) {
+      fileIDs[j] = fileID;
+    } else {
+      // We have to manually append '==' since we use the FileID string for
+      // comparing / looking up the FileID strings in the ES indices, which have
+      // the '==' appended.
+      //
+      // We may want to remove '==' in the future to reduce the uncompressed
+      // storage size by 10%.
+      fileIDs[j] = frameID.slice(0, BASE64_FILE_ID_LENGTH) + '==';
```

Comment:
I think this is wrong. You can use `frameID.slice(0, BASE64_FILE_ID_LENGTH)` for the cache lookup, but it is not exactly the `fileID` ! It is some bits more (fileID is 16 bytes, that makes 16 * 4 / 3 = 21.33 bytes base64 encoded. taking 22 bytes adds 4 bits from `addressOrLineno`.)

And this may explain my observation in the logs:
Requesting 10 minutes of data ingested locally, I get these log lines:
```
[2022-06-01T11:49:56.047+02:00][INFO ][plugins.profiling] sampleCountFromPow6 0
[2022-06-01T11:49:56.047+02:00][INFO ][plugins.profiling] query to find downsampled index took 10ms
[2022-06-01T11:49:56.173+02:00][INFO ][plugins.profiling] query to fetch events from profiling-events-all took 126ms
[2022-06-01T11:49:56.177+02:00][INFO ][plugins.profiling] processing events data took 3ms
[2022-06-01T11:49:56.178+02:00][INFO ][plugins.profiling] events total count: 15661
[2022-06-01T11:49:56.178+02:00][INFO ][plugins.profiling] unique stacktraces: 13305
[2022-06-01T11:49:56.553+02:00][INFO ][plugins.profiling] mget query (1 parallel) for 13305 stacktraces took 374ms
[2022-06-01T11:49:56.773+02:00][INFO ][plugins.profiling] processing data took 220ms
[2022-06-01T11:49:56.774+02:00][INFO ][plugins.profiling] Average size of stacktrace: 26.665631626779764
[2022-06-01T11:49:56.774+02:00][INFO ][plugins.profiling] failed to find 101 stacktraces (todo: find out why)
[2022-06-01T11:49:56.891+02:00][INFO ][plugins.profiling] mget query for 5260 executables took 110ms
[2022-06-01T11:49:56.893+02:00][INFO ][plugins.profiling] processing data took 2ms
[2022-06-01T11:49:57.011+02:00][INFO ][plugins.profiling] mget query for 31991 stackframes took 237ms
[2022-06-01T11:49:57.030+02:00][INFO ][plugins.profiling] processing data took 19ms
[2022-06-01T11:49:57.031+02:00][INFO ][plugins.profiling] found 9920 / 31991 frames
[2022-06-01T11:49:57.031+02:00][INFO ][plugins.profiling] returning payload response to client
```
Please see that the number of executables requested via mget is far too high.

Reply:
Please don't forget to amend the `Step 1:` comment as well.

Reply:
Thanks, I added your suggestion, renamed the cache, and fixed the comment in the first step.